File: modules/zig/src/main/java/com/falsepattern/zigbrains/zig/comments/ZigCommenter.java
/*
 * Copyright 2023 Mario Arias
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zig.comments;

import com.falsepattern.zigbrains.zig.psi.ZigTypes;
import com.intellij.codeInsight.generation.IndentedCommenter;
import com.intellij.lang.CodeDocumentationAwareCommenter;
import com.intellij.psi.PsiComment;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.Nullable;

public class ZigCommenter implements CodeDocumentationAwareCommenter, IndentedCommenter {

    public static final String COMMENT = "// ";
    public static final String DOC_COMMENT = "/// ";

    @Override
    public @Nullable Boolean forceIndentedLineComment() {
        return true;
    }

    @Override
    public @Nullable IElementType getLineCommentTokenType() {
        return ZigTypes.LINE_COMMENT;
    }

    @Override
    public @Nullable IElementType getBlockCommentTokenType() {
        return null;
    }

    @Override
    public @Nullable IElementType getDocumentationCommentTokenType() {
        return ZigTypes.DOC_COMMENT;
    }

    @Override
    public @Nullable String getDocumentationCommentPrefix() {
        return null;
    }

    @Override
    public @Nullable String getDocumentationCommentLinePrefix() {
        return DOC_COMMENT;
    }

    @Override
    public @Nullable String getDocumentationCommentSuffix() {
        return null;
    }

    @Override
    public boolean isDocumentationComment(PsiComment element) {
        return false;
    }

    @Override
    public @Nullable String getLineCommentPrefix() {
        return COMMENT;
    }

    @Override
    public @Nullable String getBlockCommentPrefix() {
        return null;
    }

    @Override
    public @Nullable String getBlockCommentSuffix() {
        return null;
    }

    @Override
    public @Nullable String getCommentedBlockCommentPrefix() {
        return null;
    }

    @Override
    public @Nullable String getCommentedBlockCommentSuffix() {
        return null;
    }
}

File: modules/zig/src/main/kotlin/com/falsepattern/zigbrains/zig/comments/ZigCommenter.kt
package com.falsepattern.zigbrains.zig.comments

import com.falsepattern.zigbrains.zig.psi.ZigTypes
import com.intellij.codeInsight.generation.IndentedCommenter
import com.intellij.lang.CodeDocumentationAwareCommenter
import com.intellij.psi.PsiComment
import com.intellij.psi.tree.IElementType

class ZigCommenter: CodeDocumentationAwareCommenter, IndentedCommenter {
    override fun getLineCommentPrefix() = COMMENT

    override fun getBlockCommentPrefix() = null

    override fun getBlockCommentSuffix() = null

    override fun getCommentedBlockCommentPrefix() = null

    override fun getCommentedBlockCommentSuffix() = null

    override fun forceIndentedLineComment() = true

    override fun getLineCommentTokenType() = ZigTypes.LINE_COMMENT!!

    override fun getBlockCommentTokenType() = null

    override fun getDocumentationCommentTokenType() = ZigTypes.DOC_COMMENT!!

    override fun getDocumentationCommentPrefix() = null

    override fun getDocumentationCommentLinePrefix() = DOC_COMMENT

    override fun getDocumentationCommentSuffix() = null

    override fun isDocumentationComment(element: PsiComment?): Boolean {
        val type = element?.tokenType ?: return false
        return type == ZigTypes.DOC_COMMENT || type == ZigTypes.CONTAINER_DOC_COMMENT
    }
}

const val COMMENT = "// "
const val DOC_COMMENT = "/// "